core/filterconstraint.cpp

```diff
@@ -0,0 +1,939 @@
+		return 0;
+	}
+	return c.data.timestamp_range.to;
+}
+
+uint64_t filter_constraint_get_multiple_choice(const filter_constraint &c)
+{
+	if (!filter_constraint_is_multiple_choice(c.type)) {
+		fprintf(stderr, "Getting multiple-choice of non-multiple choice constraint!\n");
+		return 0;
+	}
+	return c.data.multiple_choice;
+}
+
+// Pointer to function that takes two strings and returns whether
+// the first matches the second according to a criterion (substring, starts-with, exact).
+using StrCheck = bool (*) (const QString &s1, const QString &s2);
+
+// Check if a string-list contains at least one string containing the second argument.
+// Comparison is non case sensitive and removes white space.
+static bool listContainsSuperstring(const QStringList &list, const QString &s, StrCheck strchk)
+{
+	return std::any_of(list.begin(), list.end(), [&s,strchk](const QString &s2)
+			   { return strchk(s2, s); } );
+}
+
+// Check whether any of the items of the first list is in the second list as a super string.
+// The mode is controlled by the second argument
+static bool check(const filter_constraint &c, const QStringList &list)
+{
+	StrCheck strchk =
+		c.string_mode == FILTER_CONSTRAINT_SUBSTRING ?
+			[](const QString &s1, const QString &s2) { return s1.contains(s2, Qt::CaseInsensitive); } :
+		c.string_mode == FILTER_CONSTRAINT_STARTS_WITH ?
+			[](const QString &s1, const QString &s2) { return s1.startsWith(s2, Qt::CaseInsensitive); } :
+		/* FILTER_CONSTRAINT_EXACT */
+			[](const QString &s1, const QString &s2) { return s1.compare(s2, Qt::CaseInsensitive) == 0; };
+	return std::any_of(c.data.string_list->begin(), c.data.string_list->end(),
+			   [&list, strchk](const QString &item)
+			   { return listContainsSuperstring(list, item, strchk); }) != c.negate;
+}
+
+static bool has_tags(const filter_constraint &c, const struct dive *d)
+{
+	QStringList dive_tags;
+	for (const tag_entry *tag = d->tag_list; tag; tag = tag->next)
+		dive_tags.push_back(QString(tag->tag->name).trimmed());
+	dive_tags.append(gettextFromC::tr(divemode_text_ui[d->dc.divemode]).trimmed());
+	return check(c, dive_tags);
+}
+
+static bool has_people(const filter_constraint &c, const struct dive *d)
+{
+	QStringList dive_people;
+	for (const QString &s: QString(d->buddy).split(",", QString::SkipEmptyParts))
+		dive_people.push_back(s.trimmed());
+	for (const QString &s: QString(d->divemaster).split(",", QString::SkipEmptyParts))
+		dive_people.push_back(s.trimmed());
+	return check(c, dive_people);
+}
+
+static bool has_locations(const filter_constraint &c, const struct dive *d)
+{
+	QStringList diveLocations;
+	if (d->divetrip)
+		diveLocations.push_back(QString(d->divetrip->location).trimmed());
+
+	if (d->dive_site)
+		diveLocations.push_back(QString(d->dive_site->name).trimmed());
+
+	return check(c, diveLocations);
+}
+
+// TODO: Finish this implementation.
+/*
+static bool has_equipment(const filter_constraint &, int, const struct dive *)
+{
+	return true;
+}
+*/
```

Comment:
yeah - all my dives with twin tanks... and more importantly, all the ones without (which will bring my overall SAC rate down considerably) :-)

Reply:
Note that this was a copy&paste artifact from the old filter code. Since this is now easy to do - I implemented filtering for weight and cylinder types. We could also do cylinder-volume, pressure, etc., but for now that's not a pressing issue.